File: +tom/Optimizer.m
classdef Optimizer < handle

  methods (Access = private, Static = true)
    function dL = pDescriptionList(name, cD)
      persistent descriptionList
      if(isempty(descriptionList))
        descriptionList = containers.Map;
      end
      if(nargin==2)
        descriptionList(name) = cD;
      else
        dL = descriptionList;
      end
    end

    function fL = pFactoryList(name, cF)
      persistent factoryList
      if(isempty(factoryList))
        factoryList = containers.Map;
      end
      if(nargin==2)
        factoryList(name) = cF;
      else
        fL = factoryList;
      end
    end
  end
  
  methods (Access = protected, Static = true)
    function this = Optimizer()
    end
    
    function connect(name, cD, cF)
      if(isa(cD, 'function_handle')&&...
         isa(cF, 'function_handle'))
         tom.Optimizer.pDescriptionList(name, cD);
         tom.Optimizer.pFactoryList(name, cF);
      end
    end
  end
      
  methods (Access = public, Static = true)
    function flag = isConnected(name)
      flag = false;
      className = [name, '.', name(find(['.', name]=='.', 1, 'last'):end)];
      if(exist(className, 'class'))
        try
          feval([className, '.initialize'], name);
        catch err
          err.message;
        end  
        if(isKey(tom.Optimizer.pFactoryList(name), name))
          flag = true;
        end
      end
    end
    
    function text = description(name)
      text = '';
      if(tom.Optimizer.isConnected(name))
        dL = tom.Optimizer.pDescriptionList(name);
        text = feval(dL(name));
      end
    end
    
    function obj = create(name)
      if(tom.Optimizer.isConnected(name))
        cF = tom.Optimizer.pFactoryList(name);
        obj = feval(cF(name));
        assert(isa(obj, 'tom.Optimizer'));
      else
        error('The requested component is not connected');
      end
    end
  end
  
  methods (Abstract = true, Access = public, Static = true)
    initialize(name);
  end
  
  methods (Abstract = true, Access = public)
    num = numInitialConditions(this);
    defineProblem(dynamicModel, measure, randomize);
    refreshProblem(this);
    num = numSolutions(this);
    xEst = getSolution(this, k);
    cEst = getCost(this, k);
    step(this);
  end
  
end

File: +tom/Optimizer.py
# Copyright 2011 Scientific Systems Company Inc., New BSD License

import Measure, Trajectory, DynamicModel

from abc import abstractmethod, ABCMeta

class Optimizer(object):

    staticDescriptionList = {}
    
    @staticmethod
    def pDescriptionList():
        return Optimizer.staticDescriptionList

    staticFactoryList = {}

    @staticmethod
    def pFactoryList():
        return Optimizer.staticFactoryList

    def __init__(self):
        pass

    @staticmethod
    def connect(name, cD, cF):
        Optimizer.pDescriptionList()[name] = cD
        Optimizer.pFactoryList()[name] = cF

    @staticmethod
    def isConnected(name):
        return Optimizer.pFactoryList().has_key(name)
    
    @staticmethod
    def description(name):
        return Optimizer.pDescriptionList().get(name)

    @staticmethod
    def create(name, initialTime, uri):
        if Optimizer.isConnected(name):
            return Optimizer.pFactoryList().get(name)()
        else:
            raise ValueError('Optimizer is not connected to the requested component')

    @staticmethod
    def initialize(name):
        pass

    @abstractmethod
    def numInitialConditions(self):
        pass

    @abstractmethod
    def defineProblem(self, dynamicModel, measure, randomize):
        pass

    @abstractmethod
    def refreshProblem(self):
        pass

    @abstractmethod
    def numSolutions(self):
        pass

    @abstractmethod
    def getSolution(self, k):
        pass

    @abstractmethod
    def getCost(self, k):
        pass

    @abstractmethod
    def step(self):
        pass

